plugins/gatsby-plugin-auto-index-pages/gatsby-node.js

```diff
@@ -0,0 +1,125 @@
+const path = require('path');
+const fromList = require('./utils/unist-fs-util-from-list');
+const visit = require('unist-util-visit');
+const generateHTML = require('./utils/generate-html');
+const { sentenceCase } = require('./utils/string');
+
+exports.createPages = async ({ actions, graphql, reporter }, pluginOptions) => {
+  const { skippedDirectories } = pluginOptions;
+  const { createPage } = actions;
+
+  const { data, errors } = await graphql(`
+    query MyQuery {
+      tableOfContents: allFile(
+        filter: {
+          sourceInstanceName: { eq: "markdown-pages" }
+          base: { in: ["index.mdx", "index.md"] }
+        }
+      ) {
+        nodes {
+          childMdx {
+            fields {
+              slug
+            }
+          }
+          childMarkdownRemark {
+            fields {
+              slug
+            }
+          }
+        }
+      }
+      allFile(
+        sort: { fields: [relativePath] }
+        filter: {
+          sourceInstanceName: { eq: "markdown-pages" }
+          base: { nin: ["index.mdx", "index.md"] }
+          children: {
+            elemMatch: { internal: { type: { in: ["MarkdownRemark", "Mdx"] } } }
+          }
+        }
+      ) {
+        nodes {
+          base
+          relativePath
+          childMdx {
+            frontmatter {
+              title
+            }
+            fields {
+              slug
+            }
+          }
+          childMarkdownRemark {
+            frontmatter {
+              title
+            }
+            fields {
+              slug
+            }
+          }
+        }
+      }
+    }
+  `);
+
+  if (errors) {
+    reporter.panicOnBuild(errors[0]);
+  }
+
+  const {
+    tableOfContents: { nodes: tableOfContentsNodes },
+    allFile: { nodes: fileNodes },
+  } = data;
+
+  const existingPaths = tableOfContentsNodes
+    .map(getSlug)
+    .concat(fileNodes.map(getSlug));
+
+  const files = fileNodes.map((node) => ({
+    basename: node.base,
+    path: node.relativePath,
+    contents: node.childMdx || node.childMarkdownRemark,
+  }));
+
+  const list = fromList(files, ({ contents }) => contents);
+
+  visit(list, 'directory', (dir) => {
+    const slug = `/${dir.path}`;
+
+    if (skippedDirectories.includes(dir.path)) {
+      return [visit.SKIP];
+    }
+
+    if (existingPaths.includes(slug)) {
+      return;
+    }
+
+    createPage({
+      path: slug,
+      component: path.resolve('src/templates/indexPage.js'),
+      context: {
+        slug,
+        html: generateHTML(dir),
```

Comment:
why are we generating html for the table of contents? why isn't it in mdx?

Reply:
In this case, we aren't actually generating a file, but rather the contents of the index page in memory (errr, rather `pageContext`). MDX is really just a text format that will compile to JSX code, which is then evaluated at runtime by the `MDXRenderer`. Since its just a list of headers and links, I don't really need to compile the body to JSX and instead can go right to HTML.

This plugin is a replacement for the index pages, like this one: https://docs-preview.newrelic.com/docs/integrations

Notice how its just a list of headers and links. By compiling to HTML, I don't need to do any fancy logic in the template since the HTML is already built for me.

Reply:
thank you!! this is helpful to know :)